core/src/coproduct.rs

```diff
@@ -144,28 +144,360 @@ macro_rules! Coprod {
+impl<Head, Tail> Coproduct<Head, Tail> {
+    /// Instantiate a coproduct from an element.
+    ///
+    /// This is generally much nicer than nested usage of `Coproduct::{Inl,Inr}`.
+    /// It uses a trick with type inference to automatically build the correct variant
+    /// according to the input type.
+    ///
+    /// In standard usage, the Index type parameter can be ignored,
+    /// as it will typically be solved for using type inference.
+    ///
+    /// # Rules
+    ///
+    /// If the type does not appear in the coproduct, the conversion is forbidden.
+    /// If the type appears multiple times in the coproduct, type inference will fail.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # #[macro_use] extern crate frunk_core;
+    /// # fn main() {
+    /// use frunk_core::coproduct::Coproduct;
+    ///
+    /// type I32F32 = Coprod!(i32, f32);
+    ///
+    /// // Constructing coproducts using inject:
+    /// let co1_nice: I32F32 = Coproduct::inject(1i32);
+    /// let co2_nice: I32F32 = Coproduct::inject(42f32);
+    ///
+    /// // Compare this to the "hard way":
+    /// let co1_ugly: I32F32 = Coproduct::Inl(1i32);
+    /// let co2_ugly: I32F32 = Coproduct::Inr(Coproduct::Inl(42f32));
+    ///
+    /// assert_eq!(co1_nice, co1_ugly);
+    /// assert_eq!(co2_nice, co2_ugly);
+    ///
+    /// // Feel free to use `inject` on a type alias, or even directly on the
+    /// // `Coprod!` macro. (the latter requires wrapping the type in `<>`)
+    /// let _ = I32F32::inject(42f32);
+    /// let _ = <Coprod!(i32, f32)>::inject(42f32);
+    ///
+    /// // You can also use a turbofish to specify the type of the input when
+    /// // it is ambiguous (e.g. an empty `vec![]`).
+    /// // The Index parameter should be left as `_`.
+    /// type Vi32Vf32 = Coprod!(Vec<i32>, Vec<f32>);
+    /// let _: Vi32Vf32 = Coproduct::inject::<Vec<i32>, _>(vec![]);
+    /// # }
+    /// ```
+    #[inline(always)]
+    pub fn inject<T, Index>(to_insert: T) -> Self
+    where Self: CoprodInjector<T, Index>,
+    {
+        CoprodInjector::inject(to_insert)
+    }
+
+    /// Borrow an element from a coproduct by type.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # #[macro_use] extern crate frunk_core;
+    /// # fn main() {
+    /// type I32F32 = Coprod!(i32, f32);
+    ///
+    /// // You can let type inference find the desired type:
+    /// let co1 = I32F32::inject(42f32);
+    /// let co1_as_i32: Option<&i32> = co1.get();
+    /// let co1_as_f32: Option<&f32> = co1.get();
+    /// assert_eq!(co1_as_i32, None);
+    /// assert_eq!(co1_as_f32, Some(&42f32));
+    ///
+    /// // You can also use turbofish syntax to specify the type.
+    /// // The Index parameter should be left as `_`.
+    /// let co2 = I32F32::inject(1i32);
+    /// assert_eq!(co2.get::<i32, _>(), Some(&1));
+    /// assert_eq!(co2.get::<f32, _>(), None);
+    /// # }
+    /// ```
+    #[inline(always)]
+    pub fn get<S, Index>(&self) -> Option<&S>
```

Comment:
A thought that occurred to me just now...
Should we generalize this to `-> R` where `R: Try<&S, E>` once `Try` is stable?
(This applies generally where we have `Option<?>`)

Reply:
Eh, I'm not sure if I like it. Where does the `E` come from?  ISTM that E would have to `impl Default` or `impl From<NoneError>` or something like that and I'm not sure if there's an obvious "right choice"

Reply:
Also it would make it difficult to do things like `self.get().expect()` or `self.get().map(f)`

Reply:
I think `impl Default` would be reasonable, but I hear you on `self.get().map(f)` -- type inference is not helped.